### PR TITLE
[BREAKING][model, ops, misc] chore: upgrade transformers to 5.16.1

### PR DESCRIPTION
**Pull request by FoolPlayer**
### What does this PR do?

Upgrade the supported Transformers pin from 5.9.0 to 5.16.1 and regenerate all 29 GPU/NPU model patches. Adapt model integrations to the current upstream APIs while retaining VeOmni attention dispatch, sequence/context parallelism, QAT, and indexer-loss behavior.

DeepSeek-V4 now follows HF's `indexer.scorer.weights_proj` hierarchy. The checkpoint converter maps original inference and older VeOmni weight keys to the current model key; HF reference state dicts load strictly without a parity-test rename. Older DCP checkpoints require migration of both model and optimizer keys before resume; the safetensors converter does not migrate DCP optimizer state.

### Design & Code Changes

- Adapt Qwen vision/Omni position and feature APIs, Qwen3.5 linear attention, DeepSeek-V3 router/experts, and GLM-MoE-DSA sparse attention/cache contracts.
- Construct generated Qwen child towers explicitly so upstream AutoModel dispatch cannot bypass VeOmni patches. Keep parity reference weights on CPU between modes to fit CI accelerator memory.
- Bind Omni thinker RoPE inputs by keyword after HF added a global audio/video flag. Preserve per-video training conventions and cover silent videos, padding, and explicit global `False`.
- Carry MLA value padding into VeOmni's flash-attention wrapper and reject sparse indices unsupported by the dense wrapper.
- Use DeepSeek-V4's current scorer hierarchy on GPU/NPU and retain its upstream FP8 exclusion. Convert legacy checkpoint keys at loading/export boundaries, including safetensors shard-index mapping.
- Correct DeepSeek-V4 CP sink initialization and Qwen3.5 toy `layer_types`; measure load-balancing loss memory on the supported training-forward workload.
- Update dependency lock, installation guidance, and migration documentation. Only Transformers, tokenizers, and safetensors versions change in the lockfile. Remove unrelated documentation formatting.
- Require a user decision for model/checkpoint key conflicts in the model-support and patchgen skills; apply an existing decision without asking again.

### Test

- Key audit: inspect constructor assignments in all 29 generated modules; HF and VeOmni parameter keys match in 28 meta-device model instances. Qwen3-Omni NPU requires `torch_npu` locally and was checked statically. No other model was found to retain obsolete parameter names for compatibility. Existing MoE layout conversions remain in the checkpoint layer.
- Current scorer migration: upgrade contracts, converter, QAT, and generated-import suites pass (207 unique cases, 1 skipped for unavailable `torch_npu`), including 14 new scorer import/export regressions.
- DeepSeek-V4 strict HF/VeOmni forward/backward parity passes. Two-GPU training checkpoint save/resume and HF export pass.
- Indexer-loss and CP/Ulysses regressions: 159 passed.
- All 29 generated modules pass drift checks; `make quality`, documentation-path checks, and both edited skill validators pass. Independent review: safe.
- Earlier full-upgrade validation: all 20 remote checks passed on pre-rebase head `9cb25bb21`, including GPU/NPU unit and e2e workflows. Subsequent rebase onto `e6c46a9a3` passed 248 local checks (1 skipped, 2 xfailed). Current-head CI must run again after this update; those earlier results are not a claim about the new head.

### API and Usage Example

```bash
uv sync --extra gpu --dev
```

Pip-based installs should select `transformers==5.16.1`. See `docs/transformers_v5/upgrade_5_9_to_5_16.md` for API changes and checkpoint migration requirements.

### Checklist Before Submitting

- [x] Title follows the CI-enforced format and identifies the checkpoint key change as breaking.
- [x] Updated documentation and added regressions to existing CI-enumerated suites.
- [x] Applied quality, patchgen drift, and documentation checks.
- [x] Completed independent code review.

## Summary by CodeRabbit

* **Compatibility**
  * Updated the supported Transformers version to 5.16.1 across installation instructions, containers, and model integrations.
  * Improved compatibility across CUDA, Ascend NPU, and Cambricon MLU accelerator paths.

* **Model Support**
  * Updated DeepSeek, GLM, Qwen, and multimodal integrations for Transformers 5.16.1.
  * Improved checkpoint loading, multimodal feature handling, sparse attention, and attention-dimension compatibility.

* **Documentation**
  * Added a Transformers 5.9.0-to-5.16.1 upgrade guide and refreshed related documentation.

* **Tests**
  * Added coverage for generated model imports, multimodal model construction, and upgrade compatibility.

### CODE REVIEW

File: tests/models/test_models_patch.py [hunk 1]
```diff
@@ -269,14 +268,16 @@ def forward_backward_step(self, state_dict: Dict[str, torch.Tensor], model_mode:
...
-        # ``DeepseekV3CheckpointTensorConverter``); that path is exercised by
-        # ``test_logits_bitwise_equal_v5_via_loader`` in
-        # ``test_models_logits_equal.py``.
+        # HF 5.16 nests the V4 scoring head under ``scorer``. VeOmni keeps
+        # the original key for native model/optimizer checkpoint compatibility.
+        # Translate only this in-memory naming difference, retaining strict
+        # loading and the untouched HF snapshot for subsequent reference modes.
+        # On-disk checkpoint conversion is covered by the converter/loader tests.
+        if model_name == "deepseek_v4" and model_mode.modeling_backend == "veomni":
+            state_dict = {
```
Comment (Coach257) [inline tests/models/test_models_patch.py]: Maybe update converter instead of patching here?

**Review by Coach257 — APPROVED** [review]
